dlib/matrix/mkl_fft.h

```diff
@@ -0,0 +1,69 @@
+#ifndef DLIB_MKL_FFT_H
+#define DLIB_MKL_FFT_H
+
+#include <type_traits>
+#include <mkl_dfti.h>
+
+#define DLIB_DFTI_CHECK_STATUS(s) \
+    if((s) != 0 && !DftiErrorClass((s), DFTI_NO_ERROR)) \
+    { \
+        throw dlib::error(DftiErrorMessage((s))); \
+    }
+
+namespace dlib
+{
+    template<typename T, typename std::enable_if<std::is_floating_point<T>::value>::type* = nullptr>
+    void mkl_fft(const std::vector<int>& dims, const std::complex<T>* fin, std::complex<T>* fout, bool is_inverse)
```

Comment:
It would be good to use something other than `std::vector` for dims.   Anything that doesn't involve a memory allocation since this is just 1 or 2 numbers and an allocation is a lot of overhead for that.  I'm still looking at how this is all used, but maybe `std::array` would be good.

Reply:
Hmm, ```std::array``` is fixed size...  It means i can't do stuff like ```dims.erase(std::remove(dims.begin(), dims.end(), 1), dims.end());```

Reply:
Memory allocations are expensive.  They do something like lock a global mutex, make system calls, and all that.  If you are going to all the trouble to make the ffts faster as you are it's almost certainly avoiding memory allocations as well.

Reply:
ok cool. Will have a think. Presumably then we use std::array with a templated size? I imagine we can still pass initialiser lists as arguments?

Reply:
std::array doesn't work when passing initialiser lists. C-style arrays are fine though. I'll use that.

Reply:
The best thing would be a container that has the small size optimisation, a bit like what std::vector and std::string had in C++98 i think. So that i could do something like:
```dims.erase(std::remove(dims.begin(), dims.end(), 1), dims.end());```

Reply:
Actually, I'm gonna use std::vector for now because that's what ```plan_key``` uses. All FFT plans use the same ```plan_key``` type and i don't want to template it for different sizes of ```dims```.  I think something like [SmallVector](https://llvm.org/doxygen/SmallVector_8h_source.html) would be ideal and tick all the boxes. That could be another PR adding something like SmallVector to dlib.

Reply:
In the grand scheme of things, i don't think doing a small allocation for ```dims``` is going to be the bottleneck when computing FFTs.

Reply:
Depends on the fft size. Someone doing billions of small 1D ffts is going to care.

Reply:
For now, do you mind if i focus on getting everything working including MKL bindings, ```fftr``` getting merged into ```fft``` and adding all the overloads for normal ```matrix``` type. Little details i will do at the end, or ideally in another PR which adds performance enhancements here and there.

Reply:
Yeah that’s fine :)